File: pkg/operations/web.go
package operations

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

// WebHandler handles web automation operations
type WebHandler struct {
	client *http.Client
}

// NewWebHandler creates a new web handler
func NewWebHandler() *WebHandler {
	return &WebHandler{
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// SetTimeout sets the HTTP client timeout
func (h *WebHandler) SetTimeout(timeout time.Duration) {
	h.client.Timeout = timeout
}

// GetType returns the operation type
func (h *WebHandler) GetType() string {
	return "web"
}

// Validate checks if the parameters are valid
func (h *WebHandler) Validate(params map[string]interface{}) error {
	action, ok := params["action"].(string)
	if !ok {
		return fmt.Errorf("action parameter is required")
	}

	switch action {
	case "get", "post", "put", "patch", "delete":
		if _, ok := params["url"]; !ok {
			return fmt.Errorf("url parameter is required for %s action", action)
		}
	case "request":
		if _, ok := params["url"]; !ok {
			return fmt.Errorf("url parameter is required for request action")
		}
		if _, ok := params["method"]; !ok {
			return fmt.Errorf("method parameter is required for request action")
		}
	}

	return nil
}

// Execute performs the web operation
func (h *WebHandler) Execute(ctx context.Context, params map[string]interface{}, execCtx interface{}) (map[string]interface{}, error) {
	action := params["action"].(string)

	var result map[string]interface{}
	var err error

	switch action {
	case "get":
		result, err = h.executeGet(ctx, params)
	case "post":
		result, err = h.executePost(ctx, params)
	case "put":
		result, err = h.executePut(ctx, params)
	case "patch":
		result, err = h.executePatch(ctx, params)
	case "delete":
		result, err = h.executeDelete(ctx, params)
	case "request":
		result, err = h.executeRequest(ctx, params)
	default:
		return nil, fmt.Errorf("unsupported web action: %s", action)
	}

	return result, err
}

func (h *WebHandler) executeGet(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	url := getStringParam(params, "url")
	headers := getHeaders(params)

	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("create request failed: %v", err)
	}

	// Set headers
	for key, value := range headers {
		req.Header.Set(key, value)
	}

	startTime := time.Now()
	resp, err := h.client.Do(req)
	duration := time.Since(startTime)

	if err != nil {
		return nil, fmt.Errorf("request failed: %v", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response failed: %v", err)
	}

	return map[string]interface{}{
		"success":     resp.StatusCode >= 200 && resp.StatusCode < 300,
		"status_code": resp.StatusCode,
		"body":        string(body),
		"headers":     resp.Header,
		"duration":    duration.Milliseconds(),
	}, nil
}

func (h *WebHandler) executePost(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	url := getStringParam(params, "url")
	headers := getHeaders(params)

	var body io.Reader
	if content, ok := params["body"]; ok {
		switch v := content.(type) {
		case string:
			body = bytes.NewBufferString(v)
		case map[string]interface{}:
			jsonData, err := json.Marshal(v)
			if err != nil {
				return nil, fmt.Errorf("marshal body failed: %v", err)
			}
			body = bytes.NewBuffer(jsonData)
			if headers["Content-Type"] == "" {
				headers["Content-Type"] = "application/json"
			}
		}
	}

	req, err := http.NewRequestWithContext(ctx, "POST", url, body)
	if err != nil {
		return nil, fmt.Errorf("create request failed: %v", err)
	}

	// Set headers
	for key, value := range headers {
		req.Header.Set(key, value)
	}

	startTime := time.Now()
	resp, err := h.client.Do(req)
	duration := time.Since(startTime)

	if err != nil {
		return nil, fmt.Errorf("request failed: %v", err)
	}
	defer resp.Body.Close()

	responseBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response failed: %v", err)
	}

	return map[string]interface{}{
		"success":     resp.StatusCode >= 200 && resp.StatusCode < 300,
		"status_code": resp.StatusCode,
		"body":        string(responseBody),
		"headers":     resp.Header,
		"duration":    duration.Milliseconds(),
	}, nil
}

func (h *WebHandler) executePut(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	url := getStringParam(params, "url")
	headers := getHeaders(params)

	var body io.Reader
	if content, ok := params["body"]; ok {
		switch v := content.(type) {
		case string:
			body = bytes.NewBufferString(v)
		case map[string]interface{}:
			jsonData, err := json.Marshal(v)
			if err != nil {
				return nil, fmt.Errorf("marshal body failed: %v", err)
			}
			body = bytes.NewBuffer(jsonData)
			if headers["Content-Type"] == "" {
				headers["Content-Type"] = "application/json"
			}
		}
	}

	req, err := http.NewRequestWithContext(ctx, "PUT", url, body)
	if err != nil {
		return nil, fmt.Errorf("create request failed: %v", err)
	}

	// Set headers
	for key, value := range headers {
		req.Header.Set(key, value)
	}

	startTime := time.Now()
	resp, err := h.client.Do(req)
	duration := time.Since(startTime)

	if err != nil {
		return nil, fmt.Errorf("request failed: %v", err)
	}
	defer resp.Body.Close()

	responseBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response failed: %v", err)
	}

	return map[string]interface{}{
		"success":     resp.StatusCode >= 200 && resp.StatusCode < 300,
		"status_code": resp.StatusCode,
		"body":        string(responseBody),
		"headers":     resp.Header,
		"duration":    duration.Milliseconds(),
	}, nil
}

func (h *WebHandler) executePatch(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	url := getStringParam(params, "url")
	headers := getHeaders(params)

	var body io.Reader
	if content, ok := params["body"]; ok {
		switch v := content.(type) {
		case string:
			body = bytes.NewBufferString(v)
		case map[string]interface{}:
			jsonData, err := json.Marshal(v)
			if err != nil {
				return nil, fmt.Errorf("marshal body failed: %v", err)
			}
			body = bytes.NewBuffer(jsonData)
			if headers["Content-Type"] == "" {
				headers["Content-Type"] = "application/json"
			}
		}
	}

	req, err := http.NewRequestWithContext(ctx, "PATCH", url, body)
	if err != nil {
		return nil, fmt.Errorf("create request failed: %v", err)
	}

	// Set headers
	for key, value := range headers {
		req.Header.Set(key, value)
	}

	startTime := time.Now()
	resp, err := h.client.Do(req)
	duration := time.Since(startTime)

	if err != nil {
		return nil, fmt.Errorf("request failed: %v", err)
	}
	defer resp.Body.Close()

	responseBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response failed: %v", err)
	}

	return map[string]interface{}{
		"success":     resp.StatusCode >= 200 && resp.StatusCode < 300,
		"status_code": resp.StatusCode,
		"body":        string(responseBody),
		"headers":     resp.Header,
		"duration":    duration.Milliseconds(),
	}, nil
}

func (h *WebHandler) executeDelete(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	url := getStringParam(params, "url")
	headers := getHeaders(params)

	req, err := http.NewRequestWithContext(ctx, "DELETE", url, nil)
	if err != nil {
		return nil, fmt.Errorf("create request failed: %v", err)
	}

	// Set headers
	for key, value := range headers {
		req.Header.Set(key, value)
	}

	startTime := time.Now()
	resp, err := h.client.Do(req)
	duration := time.Since(startTime)

	if err != nil {
		return nil, fmt.Errorf("request failed: %v", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response failed: %v", err)
	}

	return map[string]interface{}{
		"success":     resp.StatusCode >= 200 && resp.StatusCode < 300,
		"status_code": resp.StatusCode,
		"body":        string(body),
		"headers":     resp.Header,
		"duration":    duration.Milliseconds(),
	}, nil
}

func (h *WebHandler) executeRequest(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	url := getStringParam(params, "url")
	method := getStringParam(params, "method")
	headers := getHeaders(params)

	var body io.Reader
	if content, ok := params["body"]; ok {
		switch v := content.(type) {
		case string:
			body = bytes.NewBufferString(v)
		case map[string]interface{}:
			jsonData, err := json.Marshal(v)
			if err != nil {
				return nil, fmt.Errorf("marshal body failed: %v", err)
			}
			body = bytes.NewBuffer(jsonData)
			if headers["Content-Type"] == "" {
				headers["Content-Type"] = "application/json"
			}
		}
	}

	req, err := http.NewRequestWithContext(ctx, method, url, body)
	if err != nil {
		return nil, fmt.Errorf("create request failed: %v", err)
	}

	// Set headers
	for key, value := range headers {
		req.Header.Set(key, value)
	}

	startTime := time.Now()
	resp, err := h.client.Do(req)
	duration := time.Since(startTime)

	if err != nil {
		return nil, fmt.Errorf("request failed: %v", err)
	}
	defer resp.Body.Close()

	responseBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response failed: %v", err)
	}

	return map[string]interface{}{
		"success":     resp.StatusCode >= 200 && resp.StatusCode < 300,
		"status_code": resp.StatusCode,
		"body":        string(responseBody),
		"headers":     resp.Header,
		"duration":    duration.Milliseconds(),
	}, nil
}

// Helper functions
func getHeaders(params map[string]interface{}) map[string]string {
	headers := make(map[string]string)
	if h, ok := params["headers"].(map[string]interface{}); ok {
		for k, v := range h {
			if str, ok := v.(string); ok {
				headers[k] = str
			}
		}
	}
	return headers
}

File: src/operations/web.py
"""
Web operations handler for the auto-workflow system.
"""
import json
import time
from typing import Dict, Any
import requests
from .handler import Handler
from ..models import ExecutionContext


class WebHandler(Handler):
    """Handles web automation operations."""

    def __init__(self):
        self.client = requests.Session()
        self.client.timeout = 30

    def set_timeout(self, timeout: int) -> None:
        """Set the HTTP client timeout."""
        self.client.timeout = timeout

    def get_type(self) -> str:
        """Return the operation type."""
        return "web"

    def validate(self, params: Dict[str, Any]) -> None:
        """Check if the parameters are valid."""
        action = params.get("action")
        if not action or not isinstance(action, str):
            raise ValueError("action parameter is required")

        if action in ["get", "post", "put", "patch", "delete"]:
            if "url" not in params:
                raise ValueError(f"url parameter is required for {action} action")
        elif action == "request":
            if "url" not in params:
                raise ValueError("url parameter is required for request action")
            if "method" not in params:
                raise ValueError("method parameter is required for request action")

    def execute(self, params: Dict[str, Any], exec_ctx: ExecutionContext) -> Dict[str, Any]:
        """Perform the web operation."""
        action = params["action"]

        if action == "get":
            return self._execute_get(params)
        elif action == "post":
            return self._execute_post(params)
        elif action == "put":
            return self._execute_put(params)
        elif action == "patch":
            return self._execute_patch(params)
        elif action == "delete":
            return self._execute_delete(params)
        elif action == "request":
            return self._execute_request(params)
        else:
            raise ValueError(f"unsupported web action: {action}")

    def _execute_get(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute GET request."""
        url = self._get_string_param(params, "url")
        headers = self._get_headers(params)

        start_time = time.time()
        response = self.client.get(url, headers=headers)
        duration = int((time.time() - start_time) * 1000)

        return {
            "success": 200 <= response.status_code < 300,
            "status_code": response.status_code,
            "body": response.text,
            "headers": dict(response.headers),
            "duration": duration,
        }

    def _execute_post(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute POST request."""
        url = self._get_string_param(params, "url")
        headers = self._get_headers(params)

        # Prepare body
        body = None
        if "body" in params:
            body = params["body"]
            if isinstance(body, dict):
                if headers.get("Content-Type") is None:
                    headers["Content-Type"] = "application/json"
                body = json.dumps(body)
            elif isinstance(body, str):
                pass
            else:
                body = str(body)

        start_time = time.time()
        response = self.client.post(url, headers=headers, data=body)
        duration = int((time.time() - start_time) * 1000)

        return {
            "success": 200 <= response.status_code < 300,
            "status_code": response.status_code,
            "body": response.text,
            "headers": dict(response.headers),
            "duration": duration,
        }

    def _execute_put(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute PUT request."""
        url = self._get_string_param(params, "url")
        headers = self._get_headers(params)

        # Prepare body
        body = None
        if "body" in params:
            body = params["body"]
            if isinstance(body, dict):
                if headers.get("Content-Type") is None:
                    headers["Content-Type"] = "application/json"
                body = json.dumps(body)
            elif isinstance(body, str):
                pass
            else:
                body = str(body)

        start_time = time.time()
        response = self.client.put(url, headers=headers, data=body)
        duration = int((time.time() - start_time) * 1000)

        return {
            "success": 200 <= response.status_code < 300,
            "status_code": response.status_code,
            "body": response.text,
            "headers": dict(response.headers),
            "duration": duration,
        }

    def _execute_patch(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute PATCH request."""
        url = self._get_string_param(params, "url")
        headers = self._get_headers(params)

        # Prepare body
        body = None
        if "body" in params:
            body = params["body"]
            if isinstance(body, dict):
                if headers.get("Content-Type") is None:
                    headers["Content-Type"] = "application/json"
                body = json.dumps(body)
            elif isinstance(body, str):
                pass
            else:
                body = str(body)

        start_time = time.time()
        response = self.client.patch(url, headers=headers, data=body)
        duration = int((time.time() - start_time) * 1000)

        return {
            "success": 200 <= response.status_code < 300,
            "status_code": response.status_code,
            "body": response.text,
            "headers": dict(response.headers),
            "duration": duration,
        }

    def _execute_delete(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute DELETE request."""
        url = self._get_string_param(params, "url")
        headers = self._get_headers(params)

        start_time = time.time()
        response = self.client.delete(url, headers=headers)
        duration = int((time.time() - start_time) * 1000)

        return {
            "success": 200 <= response.status_code < 300,
            "status_code": response.status_code,
            "body": response.text,
            "headers": dict(response.headers),
            "duration": duration,
        }

    def _execute_request(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute custom HTTP request."""
        url = self._get_string_param(params, "url")
        method = self._get_string_param(params, "method").upper()
        headers = self._get_headers(params)

        # Prepare body
        body = None
        if "body" in params:
            body = params["body"]
            if isinstance(body, dict):
                if headers.get("Content-Type") is None:
                    headers["Content-Type"] = "application/json"
                body = json.dumps(body)
            elif isinstance(body, str):
                pass
            else:
                body = str(body)

        start_time = time.time()
        response = self.client.request(method, url, headers=headers, data=body)
        duration = int((time.time() - start_time) * 1000)

        return {
            "success": 200 <= response.status_code < 300,
            "status_code": response.status_code,
            "body": response.text,
            "headers": dict(response.headers),
            "duration": duration,
        }

    def _get_headers(self, params: Dict[str, Any]) -> Dict[str, str]:
        """Get headers from parameters."""
        headers = {}
        if "headers" in params and isinstance(params["headers"], dict):
            for k, v in params["headers"].items():
                if isinstance(v, str):
                    headers[k] = v
                else:
                    headers[k] = str(v)
        return headers

    # Helper methods
    def _get_string_param(self, params: Dict[str, Any], key: str, default: str = "") -> str:
        """Get string parameter."""
        value = params.get(key)
        if isinstance(value, str):
            return value
        return str(value) if value is not None else default
